### PR TITLE
Model zoo browser

### PR DESCRIPTION
**Pull request by ebamberg**
## Description ##

This is more a Proof-of-concept.

This push adds a web-interface based on netty, react and material-ui  as frontend technology to provide simple web interface to browse models in Model-Zoo.

repository

This is a first POC-draft just to bring in the idea and first skeleton-implementation, requesting if this is of interest.

> Netty server implementation
>> publishing of static resources is a proof of concept and should be refactored
>> bundling static resources and backend into 1 jar and publish static-resource from classpath-resources inside of the jar
 

- If this change is a backward incompatible change, why must this change be made?
- Interesting edge cases to note here

### CODE REVIEW

**Comment by lanking520** [comment]
This looks nice!

**Comment by zachgk** [comment]
This is definitely of interest! In fact, we have actually been wanting to build some kind of model zoo web interface for a while now. I talked about some of the pieces I imagined for it in [this RFC](https://github.com/awslabs/djl/discussions/516).

Two small fixes:
- Should the src/main/resources/static/dist/modelzoo-frontend.js be part of a .gitignore file?
- You probably want to remove the src/test/java placeholder because we check for the presence of the directory when merging test coverage data and having the placeholder is causing it to fail the CI

**Comment by ebamberg** [comment]
> This is definitely of interest! In fact, we have actually been wanting to build some kind of model zoo web interface for a while now. I talked about some of the pieces I imagined for it in [this RFC](https://github.com/awslabs/djl/discussions/516).
> 
> Two small fixes:
> 
> * Should the src/main/resources/static/dist/modelzoo-frontend.js be part of a .gitignore file?
> * You probably want to remove the src/test/java placeholder because we check for the presence of the directory when merging test coverage data and having the placeholder is causing it to fail the CI

I added to 2 small fixes and pushed them with the commit https://github.com/awslabs/djl/pull/525/commits/5743293f4600af46b07828b77f59a37b7bf2b59a

thanks for the link to the RFC. yes that's very interesting and exactly what I have in mind. I will think about the RFC a bit over the weekend.
2 thoughts to this after first quick reading, 

1) additional to what you mentioned a web interface is not just beneficial for not so experienced user but also for enterprises who wants to host a  ModelZoo in their own network behind a proxy/firewall. Maybe mirroring the official one and add own models. So different teams can search this ModelZoo like for a NLP Model (Bert as an example) trained on domain specific language.

2) effort for the web-interface (like authentication and so on) would be significant less if using a framework like spring. But I see the benefit of not adding to much 3rd party dependencies (especially to huge frameworks like spring which a lot of transient dependencies)  and going for plain netty in this case.

File: repository/src/main/java/ai/djl/repository/responseencoder/GSONApplicationSerializer.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+import java.lang.reflect.Type;
+
+/**
+ * custom gson serializer to serialize application object.
+ *
+ * @author erik.bamberg@web.de
+ */
+public class GSONApplicationSerializer
```
[the remark this answers is not in the document]
Comment (frankfliu) [reply repository/src/main/java/ai/djl/repository/responseencoder/GSONApplicationSerializer.java:29]: This file actually is not been used, Let's remove this file.

File: repository/README.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# DJL - Repository Server
```
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/README.md:1]: Yeah I see I have cyclic dependencies now between repository and repository :-)

I guess DJL central is a good idea cause everyone knows MavenCentral and so this give an idea that DJL central have a similar purpose.

I will rename the project to "central" then cause DJL is redundant

File: repository/src/main/java/ai/djl/repository/HttpStaticFileServerInitializer.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import io.netty.handler.stream.ChunkedWriteHandler;
+
+/**
+ * initializer to setup netty instance.
+ *
+ * @author erik.bamberg@web.de
+ */
+public class HttpStaticFileServerInitializer extends ChannelInitializer<SocketChannel> {
+
+    private final SslContext sslCtx;
```
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/src/main/java/ai/djl/repository/HttpStaticFileServerInitializer.java:32]: totally agreed with this. I also don't like final members. I removed this

File: repository/src/main/java/ai/djl/repository/ModelZooRepositoryServer.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/src/main/java/ai/djl/repository/ModelZooRepositoryServer.java]: ok I have seen then you already changed this

File: repository/src/main/java/ai/djl/repository/responseencoder/JsonResponse.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * send a response to the client.
+     *
+     * @param ctx channel context
+     * @param request full request
+     * @param entity the response
+     */
+    public void send(ChannelHandlerContext ctx, FullHttpRequest request, Object entity) {
+        Gson gson = new Gson();
```
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/src/main/java/ai/djl/repository/responseencoder/JsonResponse.java]: looks much more better then my  code :-)

File: repository/src/main/java/ai/djl/repository/responseencoder/JsonResponse.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+     * If Keep-Alive is disabled, attaches "Connection: close" header to the response and closes the
+     * connection after the response being sent.
+     *
+     * @param ctx context
+     * @param request full request
+     * @param response full response
+     */
+    private void sendAndCleanupConnection(
+            ChannelHandlerContext ctx, FullHttpRequest request, FullHttpResponse response) {
+        final boolean keepAlive = HttpUtil.isKeepAlive(request);
```
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/src/main/java/ai/djl/repository/responseencoder/JsonResponse.java]: agreed

File: repository/src/main/java/ai/djl/repository/responseencoder/GSONApplicationSerializer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/src/main/java/ai/djl/repository/responseencoder/GSONApplicationSerializer.java]: oh well yes it is not used anymore. I used this in an early version seems to got obsolete over the time. So yeah fine to remove this.

File: repository/src/main/java/ai/djl/repository/handler/HttpStaticFileServerHandler.java [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+ */
+public class HttpStaticFileServerHandler extends SimpleChannelInboundHandler<FullHttpRequest> {
+
+    private static final Logger logger = LoggerFactory.getLogger(HttpStaticFileServerHandler.class);
+
+    public static final String HTTP_DATE_FORMAT = "EEE, dd MMM yyyy HH:mm:ss zzz";
+    public static final String HTTP_DATE_GMT_TIMEZONE = "GMT";
+    public static final int HTTP_CACHE_SECONDS = 60;
+    private static final Pattern INSECURE_URI = Pattern.compile(".*[<>&\"].*");
+    private static final Pattern ALLOWED_FILE_NAME = Pattern.compile("[^-\\._]?[^<>&\\\"]*");
```
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/src/main/java/ai/djl/repository/handler/HttpStaticFileServerHandler.java]: ok I completely removed sendDirectory() and sendRedirect() now. so this regex is not used anymore.

to be honest I took a netty-fileserver-project to create this POC implementation , was very handy for the first tests. But we can remove all this now.

File: repository/src/main/java/ai/djl/repository/handler/HttpStaticFileServerHandler.java [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+            uri = "/src/main/resources/static/dist/" + uri;
+        }
+
+        final String path = sanitizeUri(uri);
+        if (path == null) {
+            this.sendError(ctx, HttpResponseStatus.FORBIDDEN);
+            return;
+        }
+
+        File file = new File(path);
```
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/src/main/java/ai/djl/repository/handler/HttpStaticFileServerHandler.java]: very good suggestion. I like that change

File: repository/src/main/java/ai/djl/repository/handler/HttpStaticFileServerHandler.java [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+        if (file.isHidden() || !file.exists()) {
```
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/src/main/java/ai/djl/repository/handler/HttpStaticFileServerHandler.java]: very good suggestion. I like that change

File: repository/src/main/java/ai/djl/repository/handler/HttpStaticFileServerHandler.java [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+            this.sendError(ctx, HttpResponseStatus.NOT_FOUND);
+            return;
+        }
+
+        if (file.isDirectory()) {
```
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/src/main/java/ai/djl/repository/handler/HttpStaticFileServerHandler.java]: very good suggestion. I like that change

File: repository/src/main/java/ai/djl/repository/handler/HttpStaticFileServerHandler.java [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+    }
+
+    /**
+     * Sets the content type header for the HTTP Response.
+     *
+     * @param response HTTP response
+     * @param file file to extract content type
+     */
+    private static void setContentTypeHeader(HttpResponse response, File file) {
+        MimetypesFileTypeMap mimeTypesMap = new MimetypesFileTypeMap();
```
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/src/main/java/ai/djl/repository/handler/HttpStaticFileServerHandler.java]: ok I am fine with that. Makes sense to reuse functionality from the ai.djl.api project

File: repository/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/README.md:1]: renamed package to serving.central and project to central. 

make this as a subproject with the next comment/PR

File: repository/README.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+## Overview
+
+This module contains an universal model repository web interface implementation.
```
[the remark this answers is not in the document]
Comment (ebamberg) [reply repository/README.md]: yes I think a submodule under serving would be a good place for the UI.

1) a plugin sounds good. That means you would like to extract the static-file-handler to an project like serving:serving_utils
and then we have the 2 projects serving:serving and serving:central both using serving_utils as a dependency ?

2) yes  agreed , both of the projects are going hand in hand. 

to 3) I would not mix the 2 into one in my opinion. It sounds more logical to me to have a DJL-Central server and from there you can for example download a model as a preconfigured Serving-docker-image or serving-jar. So you have a very small serving component to "run" a model as lightweight as possible. in my opinion there is no need every instance of serving to have all functionality from the UI like Searching, maybe Charts for training-metrics and so on.   

Or do you mean "Serving" is the basic server and you can add either a model-instance for serving or the UI component ?

**Review by frankfliu — APPROVED** [review]

**Review by zachgk — APPROVED** [review]

**Pull request merged**